Verify that main info response returns correct product headers

swallez opened the pull request:
Follow-up to #73434

Ensures that High Level Rest Client is running against a verified
Elasticsearch. When the first request is sent on HLRC, a request to the
info endpoint is made first to verify the product identification and
version.

swallez committed:
Verify main info response returns correct product headers

Follow-up to #73434

Ensures that High Level Rest Client is running against a verified
Elasticsearch. When the first request is send on HLRC, a request to the
info endpoint is made first to verify the product identification and
version.

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

swallez commented:
Build failures are because https://github.com/elastic/elasticsearch/pull/73434 has not been backported to `7.x` (see https://github.com/elastic/elasticsearch/pull/73434#issuecomment-856896383). The code is ready for review though.

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
         return response.getStatusLine().getStatusCode() == 200;
     }
 
+    private Cancellable performClientRequestAsync(Request request, ResponseListener listener) {
```
This took me a considerable amount of time to figure out what was actually happening in this method. Could you add some comments to help future readers determine what this method is doing?

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
+                if (validation.isEmpty()) {
+                    Cancellable call = client.performRequestAsync(request, listener);
+                    // Propagate cancellation
+                    cancelFlag.whenComplete((r, t) -> call.cancel());
```
Why do we cancel the call when the future is complete? Won't the future be completed successfully but the original call still be cancelled?

swallez replied:
Not sure I understnad the question. The `cancelFlag` future is completed successfully when the method's result is cancelled. If/when that happens, we also need to cancel the asynchronous API call request.

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
+
+                    @Override
+                    public void onFailure(Exception exception) {
+                        // Fail the requests (this one and the ones waiting for it) and clear the future
+                        // so that we retry the next time the client executes a request.
+                        versionValidationFuture = null;
+                        future.onFailure(exception);
```
I believe the plan was if the request to `/` failed with a 401 or 403, then subsequent checks must be ignored. We check for 401 and 403 in the `onSuccess` handler, but aren't those considered exceptional cases which invoke the `onFailure` response handler?

swallez replied:
Good catch, LLRC will indeed raise an exception for non-nominal response status codes. I added them to the `"ignore"` list in d981a15 along with some tests.

dakrone commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:
```diff
@@ -951,6 +1001,61 @@ public void testApiNamingConventions() throws Exception {
+        }
+    }
+
+    public void testProductCompatibilityCheck() throws Exception {
```
I think we need a check for when the root request is exceptional, throwing either a 401 or 403, and also if it throws something other than those status requests.

swallez replied:
See d981a15 for 401 & 403. Other failures of the info request are propagated as failures for the original API request. If `/` returns something other than 200, 401 or 403 then we have a bigger problem and we won't let API requests through, both because verification could not happen and also because they will most likely fail too.

dakrone reviewed (changes requested):
Thanks for working on this @swallez, I left some comments

swallez committed:
Add comments

swallez committed:
Consider 401 and 403 as nominal responses for product validation

swallez commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
+
+            @Override
+            void runIfNotCancelled(Runnable runnable) {
+                if (cancelFlag.isDone()) {
```
The `cancelFlag` future is completed (and not cancelled) when the `Cancellable` returned as a result is cancelled. This is because `CompletableFuture` allows to listen to normal and exceptional completion but not on cancellation. Also there's `isCancelled` and `isCompletedExceptionally` but no `isCompletedSuccessfully`. So `isDone` works here, since the only thing we ever do to this future is to `complete(null)`.

The comments added in e0e0638 provide more details on this flag.

swallez replied:
Actually the above comment is wrong: cancelling a `CompletableFuture` completes it exceptionnally with a `CancellationException`. I'll fix the comment, but woud like to keep the current behavior as we're not cancelling the flag but raising it.

swallez committed:
Update comment on future cancellation

dakrone reviewed (approved):
LGTM, thanks for iterating on this. I left a few other minor nits but nothing major.

jbaiera reviewed (changes requested):
A quick note about compatibility coverage that should be addressed before merge:

swallez committed:
Add build flavor and tagline to product compatibility verification

jbaiera commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -2172,7 +2172,20 @@ public void onFailure(Exception exception) {
-        if (major == 6 || (major == 7 && minor < 14)) {
+
+        if (major == 6) {
+            return Optional.empty();
```
I think we want to do the tagline validation here as well.

swallez committed:
Refactor/rename and document cancellation forwarding

swallez committed:
Fix forbidden API call

swallez committed:
Fix build flavor verification

swallez commented:
@dakrone I added some comments and refactored a bit in 115e9d1 and 94d88f7: `cancelFlag` is now a `cancellationForwarder` that is effectively cancelled.

@jbaiera I added build flavor and tagline validation in ceb527a and b87302d

Can you both please cross-check? Thanks!

dakrone commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java:
```diff
@@ -2002,6 +2016,205 @@ protected static boolean convertExistsResponse(Response response) {
+        String responseFlavor = mainResponse.getVersion().getBuildFlavor();
+        if ("default".equals(responseFlavor) == false) {
+            // Flavor is unknown when running tests, and non-mocked responses will return an unknown flavor
+            if (Build.CURRENT.flavor() != Build.Flavor.UNKNOWN || "unknown".equals(responseFlavor) == false) {
+                return Optional.of("Invalid or missing build flavor [" + responseFlavor + "]");
+            }
```
This needs to take into account `Build.Flavor.OSS`, because a 7.10 OSS build should still allow the HLRC to be used I think?

dakrone reviewed (changes requested):
I think we need to support OSS builds here also, so we need to change the build flavor check

swallez committed:
Also check tagline for version 6.x

jbaiera reviewed (approved):
LGTM! Thanks @swallez!

dakrone reviewed (approved):
LGTM also, thanks for doing this @swallez!

swallez commented:
Thanks you both for your patience! Now waiting for CI to hopefully be happy and I'll hit that merge button 😉
